Improve avatar handling

pmiossec opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Improvements over #10020

Note: a lot easier to review commit by commit.

## Proposed changes

* perf: increase a lot perf to download avatars (ex on a blame display: from 1m30s to 6s)
* perf: fix serialization that was broken for some avatars
* Always obtain the same Author initials image between 2 GitExtensions session
* return more meaningful author initials
* Better set of colors for Author initials background colors
* perf: make that cache never expire for avatar that never changes
* Perf: slight optimization of GravatarProvider
* Perf: prevent requesting the same avatar multiple times
* saner performance default values

## Screenshots <!-- Remove this section if PR does not change UI -->

For choice of colors:

### Before
| Before  |  After |
|---|---|
| ![image](https://user-images.githubusercontent.com/460196/229131809-a53290da-24c6-4c71-bde0-a76fdc2d08eb.png)  | ![image](https://user-images.githubusercontent.com/460196/229132118-f5cb5af7-0074-489b-a0b1-80335dd6da44.png) |

## Test methodology <!-- How did you ensure quality? -->

- Manual
- Some unit tests

## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 33.33.33
- Build de0dcb3112d960166620a45a5bf6e49f6918ea3b (Dirty)
- Git 2.38.1.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.14
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions

## Merge strategy

<!-- Change the following if the merge strategy should be changed:
- Squash merge (maintainer to decide merge message, PR submitter should cleanup commits/messages at PR approval).
- Rebase merge (PR submitter must change the commit message for the last commit).
- Merge commit. (PR submitter to rebase and squash before merges).
- To be decided later.
The maintainer may still request the contributor to squash and rebase, to make sure that merges and commit messages are clarified.
-->

Merge commit

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on GitUI/Avatars/GravatarProvider.cs:
```diff
@@ -14,6 +14,14 @@ public sealed class GravatarProvider : IAvatarProvider
+        #pragma warning disable SYSLIB0021 // 'MD5CryptoServiceProvider' is obsolete
+
+        // Hash specified at http://en.gravatar.com/site/implement/hash/
+        private readonly MD5CryptoServiceProvider _md5 = new();
```
The warning message gives a hint about easy-to-use replacement.

pmiossec replied:
Thanks for the head up. I just moved the code without a careful look and was thinking that it was the MD5 hashing algorithm that was obsoleted. In fact, it was just the API. I will push the fix...

pmiossec replied:
done.

gerhardol reviewed (commented):
+1
Would prefer to run this some too

mstv commented on GitUI/Avatars/AvatarMemoryCache.cs:
```diff
@@ -12,6 +14,7 @@ namespace GitUI.Avatars
     public sealed class AvatarMemoryCache : IAvatarProvider, IAvatarCacheCleaner
     {
         private readonly MruCache<(string email, int imageSize), Image> _cache;
+        private List<(string email, int imageSize)> _requested = new(6);
```
```suggestion
        private HashSet<(string email, int imageSize)> _requested = new(6);
```

pmiossec replied:
done.

mstv reviewed (commented):
Looks sensible, have not run
`var` should be replaced with explicit types

RussKie commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -65,22 +79,42 @@ private static (string? name, char[]? separator) NameSelector(string? name, stri
+                    return $"{char.ToUpper(name[0])}{name[1]}";
+                }
+
+                var splitNames = name.Split(_emailInitialSeparator);
```
[nit] avoid `var` unless the type is obvious

pmiossec replied:
done.

RussKie commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -18,12 +18,26 @@ public class InitialsAvatarProvider : IAvatarProvider
             return Task.FromResult<Image?>(avatar);
         }
 
+        private int GetDeterministicHashCode(string str)
```
It'd be good to have some comments explaining what's going on

pmiossec replied:
I wrote something....

pmiossec commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -588,10 +588,16 @@ public static bool ShowAuthorAvatarInCommitInfo
         public static AvatarProvider AvatarProvider
         {
-            get => GetEnumViaString("Appearance.AvatarProvider", AvatarProvider.Default);
+            get => GetEnumViaString("Appearance.AvatarProvider", AvatarProvider.None);
```
In fact, the Initials is the default value of `AvatarFallbackType` settings. So yes, initials will end up to be the default avatar image on a fresh install. It has the advantage to be really fast, 100% local, much more recognizable than other fallback and do not expire...

vbjay replied:
Code in c# a way to get composite color so no matter the background color a text color will show. 

To get a composite color that will contrast with any given background color, you can use the following C# code:

scssCopy code

Color GetContrastColor(Color backgroundColor) { // Calculate the luminance of the background color double luminance = (0.2126 * backgroundColor.R + 0.7152 * backgroundColor.G + 0.0722 * backgroundColor.B) / 255; // Use a threshold value to determine whether to use black or white as the foreground color Color foregroundColor = (luminance > 0.5) ? Color.Black : Color.White; return foregroundColor; } 

This method takes a Color object representing the background color and returns a Color object representing the foreground color that will provide the best contrast with the background.

The method works by calculating the luminance of the background color using the standard formula for converting RGB values to grayscale. If the luminance is greater than 0.5, it means the background color is relatively light, so the method returns black as the foreground color. If the luminance is less than or equal to 0.5, the background color is relatively dark, so the method returns white as the foreground color.

You can use this method in your code by calling GetContrastColor() with the background color you want to contrast with. For example:

scssCopy code

Color backgroundColor = Color.FromArgb(255, 128, 0); // orange Color foregroundColor = GetContrastColor(backgroundColor); Console.WriteLine($"Foreground color: {foregroundColor}"); 

This code will output "Foreground color: Color [A=255, R=255, G=255, B=255]" because white is the best foreground color to use with an orange background

gerhardol replied:
git-range-diff was very helpful in this review....

@vbjay Thanks for the response, quoted in the other comment.

vbjay replied:
No prob. Why I sent it was exactly for the idea of when theming works. The
dark theme can have a gradient of light color backgrounds to show a good
contrast on top of a dark background. Along with text color being good on
said background. Then flip the idea on light backgrounds.

I used chatgpt and thought it would be cool to just copy the original
convo.

Github email handler could be better  If replying to a review thread email, should put reply in that thread.

pmiossec commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -18,12 +18,26 @@ public class InitialsAvatarProvider : IAvatarProvider
             var (initials, hashCode) = GetInitialsAndHashCode(email, name);
 
-            var avatarColor = _avatarColors[Math.Abs(hashCode) % _avatarColors.Length];
+            var avatarColor = _avatarColors[hashCode];
```
I made this change because it was easier to debug when you have to see if things are working well. But I can revert it...

pmiossec commented on GitUI/Avatars/FileSystemAvatarCache.cs:
```diff
@@ -52,7 +52,19 @@ void WriteImage()
                 catch
                 {
-                    // do nothing
+                    // Workaround to avoid the "A generic error occurred in GDI+." exception when saving
```
I don't really explain the workaround (even if I can guarantee it is working 100% well!).
It's really difficult to have the inner problem that cause the exception. I got it once by luck but I don't remember exactly. If I remember well the resource is locked but copying it like that is working. Don't ask me why...

Even if the answers are not exactly describing our case, I saw multiple times in stackoverflow this workaround, like in this answer ( https://stackoverflow.com/a/59069870 ) and that's right that it works...

pmiossec commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -123,10 +123,11 @@ private static (string? name, char[]? separator) NameSelector(string? name, stri
-            Color.DarkOrange
+            Color.OrangeRed,
+            Color.Teal,
+            Color.OliveDrab,
```
I had a lot of difficulties to find colors that are dark enough (to be able to see initials in white) and different enough to be easily distinguished without ending up with something horrible.

But if you have a lot of time to spend on that, I will be happy to change it.
Maybe it's easy to add a hidden setting for that. I will have a look...

gerhardol replied:
I am not good at selecting colors, better at complaining...
The colors are not that different, really just the SlateGray that was added and that happen to be on quite a lot of the example...
Maybe removing SlateGray is an option, keep it for now.

gerhardol replied:
> Code in c# a way to get composite color so no matter the background color a text color will show.
>
> To get a composite color that will contrast with any given background color, you can use the following C# code:
>
> scssCopy code
>
> Color GetContrastColor(Color backgroundColor) { // Calculate the luminance of the background color double luminance = (0.2126 * backgroundColor.R + 0.7152 * backgroundColor.G + 0.0722 * backgroundColor.B) / 255; // Use a threshold value to determine whether to use black or white as the foreground color Color foregroundColor = (luminance > 0.5) ? Color.Black : Color.White; return foregroundColor; }
>
> This method takes a Color object representing the background color and returns a Color object representing the foreground color that will provide the best contrast with the background.
>
>The method works by calculating the luminance of the background color using the standard formula for converting RGB values to grayscale. If the luminance is greater than 0.5, it means the background color is relatively light, so the method returns black as the foreground color. If the luminance is less than or equal to 0.5, the background color is relatively dark, so the method returns white as the foreground color.
>
>You can use this method in your code by calling GetContrastColor() with the background color you want to contrast with. For example:
>
> scssCopy code
> 
> Color backgroundColor = Color.FromArgb(255, 128, 0); // orange Color foregroundColor = GetContrastColor(backgroundColor); Console.WriteLine($"Foreground color: {foregroundColor}");
>
>This code will output "Foreground color: Color [A=255, R=255, G=255, B=255]" because white is the best foreground color to use with an orange background

Colors are complex and require tuning. I would have expected that there were more guidelines and libs supporting this.
Still amazed by the work done by @NikolayXHD for the dark theme, so sad to loose it.
This would be even harder with another theme.

pmiossec replied:
i have pushed a commit so that there is a (hidden) setting for the colors list. Like that every one could remove the Grey 😉

pmiossec commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -79,22 +79,42 @@ private static (string? name, char[]? separator) NameSelector(string? name, stri
+
+                if (char.IsUpper(name[1]))
+                {
+                    return $"{char.ToUpper(name[0])}{name[1]}";
```
done.

vbjay commented:
No prob. Why I sent it was exactly for the idea of when theming works. The
dark theme can have a gradient of light color backgrounds to show a good
contrast on top of a dark background. Along with text color being good on
said background. Then flip the idea on light backgrounds.

I used chatgpt and thought it would be cool to just copy the original
convo.

On Sat, Apr 1, 2023, 3:58 PM Gerhard Olsson ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/10846#pullrequestreview-1367919288>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3WDMN4V2TVKE534BBFR7TW7CCG7ANCNFSM6AAAAAAWOVCCQE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mstv commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -65,35 +87,76 @@ private static (string? name, char[]? separator) NameSelector(string? name, stri
+
+            try
+            {
+                var alpha = Convert.ToInt32(c.Substring(1, 2), 16);
+                var red = Convert.ToInt32(c.Substring(3, 2), 16);
+                var green = Convert.ToInt32(c.Substring(5, 2), 16);
+                var blue = Convert.ToInt32(c.Substring(7, 2), 16);
```
use `int`?

vbjay replied:
Or use the class already made for this purpose. 

https://learn.microsoft.com/en-us/dotnet/api/system.drawing.colorconverter

vbjay replied:
Or just pass the full number instead of substring 

https://learn.microsoft.com/en-us/dotnet/api/system.drawing.color.fromargb

vbjay replied:
https://www.c-sharpcorner.com/uploadfile/mahesh/the-colorconverter-and-colortranslater-classes-in-gdi/

Basically use the right one based on what you want to store and retrieve.

RussKie replied:
@vbjay do you mean [`ColorTranslator.FromHtml(String)`](https://learn.microsoft.com/en-us/dotnet/api/system.drawing.colortranslator.fromhtml?view=net-7.0#system-drawing-colortranslator-fromhtml(system-string))?

pmiossec replied:
done.

vbjay replied:
@RussKie  yes. or the color converter way... Either way we shouldn't be rolling our own way when .net has ways to translate colors.

vbjay replied:
Once you have the color you can retrieve the parts.
```c#
Color c = Colors.Cyan;

Alpha= c.A;
Red=c.R;
```

mstv reviewed (approved):
LGTM, have not run

RussKie commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -16,27 +16,49 @@ public class InitialsAvatarProvider : IAvatarProvider
             return Task.FromResult<Image?>(avatar);
         }
 
+        /// <summary>
+        /// Calculate the most simpler non-cryptographic deterministic hash (to get the same result every times it is calculated for the same string)
+        /// We just need a not too costly way to convert a string in an integer and calculating a hash is a good way to do it.
+        /// We are avoiding to use `.GetHashCode()` because it returns a different value for each process :(
+        /// Taken from https://stackoverflow.com/a/5155015
+        /// </summary>
+        /// <param name="str">the string to calculate the hash for</param>
+        /// <returns>the hash calculated</returns>
```
```suggestion
        /// <summary>
        /// Calculate the most simpler non-cryptographic deterministic hash (to get the same result every times it is calculated for the same string)
        /// We just need an inexpensive way to convert a string to an integer, and calculating a hash is a good way to do it.
        /// We are not using <c>GetHashCode()</c> because it returns a different value for each process.
        /// Borrowed from https://stackoverflow.com/a/5155015
        /// </summary>
        /// <param name="str">The string to calculate a hash for.</param>
        /// <returns>The calculated hash.</returns>
```

RussKie commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -65,35 +87,76 @@ private static (string? name, char[]? separator) NameSelector(string? name, stri
                 {
                     // ... return that character as uppercase
-                    return names[0][0].ToString().ToUpper();
+                    return name[0].ToString().ToUpper();
```
Why not this?
```suggestion
                    return char.ToUpper(name[0]);
```

RussKie reviewed (commented):
Overall looks good. Few nits. Modulo the question from/suggestion from @vbjay.

vbjay commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -65,35 +87,72 @@ private static (string? name, char[]? separator) NameSelector(string? name, stri
+            catch
+            {
+                return Color.Black;
+            }
```
Don't make it so complicated...

![image](https://user-images.githubusercontent.com/3629489/229351373-09e84137-4cba-4b2c-963d-4d7babd1a22d.png)
![image](https://user-images.githubusercontent.com/3629489/229351391-5c08030e-ac45-40ef-828f-b3d979184199.png)
![image](https://user-images.githubusercontent.com/3629489/229351424-128f8977-1f39-45f3-bab2-4b8b5bf389eb.png)


```C#
Color named = Color.Cyan;
Color other = Color.FromArgb(named.ToArgb() + 1);

var translations = new { Named = named, NamedTranslated = ColorTranslator.ToHtml(named), Other = other, OtherTranslated = ColorTranslator.ToHtml(other) };

translations.Dump("TranslatedTo");

var BackToColors = new { Named = ColorTranslator.FromHtml(translations.NamedTranslated), Other = ColorTranslator.FromHtml(translations.OtherTranslated) };
BackToColors.Dump("RoundTrip");

```

vbjay replied:
```suggestion
             try
            {
                return ColorTranslator.FromHtml(c);
            }
            catch
            {
                return Color.Black;
            }
```

vbjay replied:
```C#
Dictionary<string, Color> colorDict = new Dictionary<string, Color>();
foreach (var prop in typeof(Color).GetProperties())
{
	if (prop.PropertyType == typeof(Color))
	{
		Color color = (Color)prop.GetValue(null);
		colorDict.Add(prop.Name, color);
	}
}

colorDict.Keys.Select(k => System.Drawing.ColorTranslator.FromHtml(k)).Dump();
```

Did not cause an exception  while 

![image](https://user-images.githubusercontent.com/3629489/229352714-903437a2-f8c7-4324-a3e0-8ec50d0aea61.png)

pmiossec replied:
I pushed changes hours ago. It seems that you are still basing your comments on the old version. Have a look...

vbjay replied:
Not needed

Stop using fromname

![image](https://user-images.githubusercontent.com/3629489/229355430-96f05732-ae02-4f1f-995c-41262edadebf.png)
As I have proven in my comments all you need is the colortranslator.

vbjay replied:
https://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/ColorConverter.cs,109

https://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/Advanced/ColorTranslator.cs,228

See how ColorTranslator handles all the names and even the LightGray name difference?

vbjay replied:
.net versions of links 

https://source.dot.net/#System.Drawing.Primitives/System/Drawing/Color.cs,475

Colors used by the abov emethod https://source.dot.net/System.Drawing.Primitives/src/libraries/Common/src/System/Drawing/ColorTable.cs.html#6532ea3b4c32baf5

While Translator is the following

https://source.dot.net/#System.Drawing.Primitives/System/Drawing/ColorTranslator.cs,218

pmiossec replied:
Indeed. Done.

vbjay commented on GitUI/Avatars/InitialsAvatarProvider.cs:
```diff
@@ -65,35 +87,60 @@ private static (string? name, char[]? separator) NameSelector(string? name, stri
-        };
+            try
+            {
+                return !colorCode.StartsWith("#") ? Color.FromName(colorCode) : ColorTranslator.FromHtml(colorCode);
```
```suggestion
                return ColorTranslator.FromHtml(colorCode);
```

vbjay replied:
See https://github.com/gitextensions/gitextensions/pull/10846/files#r1155297871 comment chain

pmiossec commented:
I think I fixed all the review comments...

vbjay commented:
Just an improvement on my contrasting colors comment

https://gist.github.com/vbjay/cc56f5e67a9165bd022418167b6af323

![image](https://user-images.githubusercontent.com/3629489/229387245-94902b0c-e427-43f0-a180-3a665770773a.png)

Open in linqpad and run.  Play with  thresholds and such to tweak.

vbjay commented:
I'd change the getconteastingcolor function use the dynamic thresholds idea. That allows us to tweak the low or high value if needed.  

This is an extension method and we could make it part of GitGui extensions. That way if something else needs contrasting color it is reusable and consistent. 


https://gist.github.com/vbjay/cc56f5e67a9165bd022418167b6af323

mstv commented:
(GH supports delta reviews: blue button "Changes since last review". I find this very useful. But GH loses track after too many rebases / force pushes. For the future, please rebase only in order to resolve conflicts or perhaps after a longer inactivity. Thank you.)

pmiossec commented:
> (GH supports delta reviews: blue button "Changes since last review". I find this very useful. But GH loses track after too many rebases / force pushes. For the future, please rebase only in order to resolve conflicts or perhaps after a longer inactivity. Thank you.)

I didn't know this option that looks great. I will keep for now on the fixup/rebase as much as possible for a last pre-merge step.

pmiossec commented:
> I'd change the getcontrastingcolor function use the dynamic thresholds idea. That allows us to tweak the low or high value if needed.
> 
> This is an extension method and we could make it part of GitGui extensions. That way if something else needs contrasting color it is reusable and consistent.
> 
> https://gist.github.com/vbjay/cc56f5e67a9165bd022418167b6af323
> 
> Could even adjust extension method to have an overload that accepts the thresholds if we wanted to override for some reason.

@vbjay Reading multiple times the code, I absolutely don't understand threshold part. Either it is very clever and complicated and I don't get it or the implementation is bugged. I tend for the 2nd option without being certain: why changing the threshold value based on the luminance and after comparing them whereas a simple comparison seems enough? 

So I have made something simpler that I think is enough...

Have a look, please.

vbjay commented:
The idea is if needed you can set <.5 threshold to .2 and >.5 to .8 and it adjusts the points where color flips to white. 


Open script in gist with linkpad.net and play with adjusting threshold values and running.

pmiossec commented:
> The idea is if needed you can set <.5 threshold to .2 and >.5 to .8 and it adjusts the points where color flips to white.

That's what my code is doing also, I think.
And playing with linqpad does not help understand the code....
I spent int total more than 20min in the 2 loc below without understanding the added value to change the threshold based on the luminance value **AND AFTER** comparing these values instead of changing the threshold value directly which is way more clearer.

Because, for me it's a lot of complexity for something wrong. Let's see with the 0.2 and 0.8 values you are mentioning, so the code:

```
	// Use a dynamic threshold based on the luminance of the background color.  Play with threshold values when <.5 or not to tweak the flip
	double threshold = (luminance < 0.5) ? 0.2 : 0.8;

	// Use the threshold value to determine whether to use black or white as the foreground color
	Color foregroundColor = (luminance > threshold) ? Color.Black : Color.White;
```

The results:

| | luminance < 0.2 |  0.2 < luminance < 0.5 | 0.5 < luminance < 0.8  | luminance > 0.8  |
|-|-----------------|---|---|---|
|    Background |   Very dark | dark  | light | Very light  |
|Threshold (updated)  | 0.2  |  0.2  |  0.8  | 0.8  |
| Foreground |     `false` => White       | `true` => Black  | `false` => White  | `true` => Black  |
| What we want? |    White on Very dark:👍           | Black on dark:👎  |  White on light:👎 | Black on Very light:👍  |

For me the only good behavior is when the 2 values are the same which is equivalent to my (a lot simpler to understand) code...

Or I misunderstood everything and that's not all the values are expected to be set? In this case, can you explain in details?

gerhardol commented:
Ready to merge?

RussKie commented:
@vbjay has approved, so I think it is.

On Sat, 8 Apr 2023, 8:44 am Gerhard Olsson, ***@***.***>
wrote:

> Ready to merge?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/10846#issuecomment-1500703169>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVLX7K4QOO7HYIEVFTXACKFFANCNFSM6AAAAAAWOVCCQE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

vbjay commented:
Correct.  I was just pushing for a little more control over the threshold and fine tuning if contrast didn't quite work.

pmiossec commented:
> Correct. I was just pushing for a little more control over the threshold and fine tuning if contrast didn't quite work.

Yes but except if I missed something, your proposition is not working or not adding value, with much more complexity, as I shown in my list comment. I was waiting for your answer on that.

If you are all OK, I could rebase and squash one or 2 commits to be ready to go...

vbjay commented:
Last attempt. Think of it instead of a single slider you have 2. If low you can independently adjust where the threshold if less than .5 or darker colors. This .5 threshold isn't perfect and allowing 2 thresholds lets us independently adjust the thresholds if light or dark.

vbjay commented:
The other point. I wanted you to keep it sn extension method. So if and when a color needs to get contrast it is c.GetContrastColor

pmiossec commented:
> Last attempt. Think of it instead of a single slider you have 2. If low you can independently adjust where the threshold if less than .5 or darker colors. This .5 threshold isn't perfect and allowing 2 thresholds lets us independently adjust the thresholds if light or dark.

I understand the general idea but I would have preferred an explanation with a real example with working threshold values because except if you found an error in the one I gave, we see that the solution with 2 threshold values is not working and add a lot of complexity.

Also, having only one setting is nearly self-explanatory if someone wants to play with it. Good luck for the explanation and the documentation if we include the 2 threshold settings... I spend quite some time investigating it and I still can't get the added value...(for me it's not working!) 

I will be interested by another point of view on the subject. 

> The other point. I wanted you to keep it an extension method. So if and when a color needs to get contrast it is c.GetContrastColor

I will have a look at that...

pmiossec commented:
> The other point. I wanted you to keep it an extension method. So if and when a color needs to get contrast it is c.GetContrastColor

Done.

I have updated the PR to merge into the new `release/4.1` branch also.
If we decide to not include in the 4.1, I will change it back....

pmiossec commented:
> @pmiossec merge at will

Thanks. Done.
